### PR TITLE
Déjà Vu Diversity Aid

### PR DESCRIPTION
**Pull request by JyeSmith**
Déjà Vu Diversity Aid (DVDA) introduces 2 new rates for discussion, D250 and D500.  Both rates run at F1000 OTA, but to the handset and flight controller the rates are 250 and 500Hz.  Double or quadruple packets are transmitted at the 1kHz OTA rate, and this aims to improve LQ compared to current L250 and L500 at close range.  Soooo much testing to do.

D250 and D500 LQ calculations do not take into account sync and tlm slots.  Even with these enabled the FC should receive a packet every slot e.g. at D250 the RC packet has 4 tx slots.  If 1 is taken up by a sync packet this still leaves 3 slots to send the RC data.

The FHSShopInterval has been reduced from 8 to 2.  This minimises blocks of packet losses that were seen at a FHSShopInterval of 8.  D250 now sends 4 packets over 3 frequencies, and D500 sends 2 packets over 2 frequencies.

### Tx Diversity
For all normal rf modes and D500 the Tx antennas will switch every packet.

For D250 which has 4 repeat packets, the antennas will switch every 2 transmissions e.g. new data from the handset > antenna switches > send OTA > FHSS > send OTA > antenna switches >  send OTA > FHSS > send OTA > new data from the handset > ... 

### Latency 
Latency and interval comparison between D250 and L250.  D250 latency is increased by about 150us, but as expected the interval is much more consistent without increases due to missed packets, tlm, or sync packets.

image

D500 and L500.  There does not appear to be a noticeable increase in latency.

image

And here is the OTA packet rate F1kHz.

image

### CODE REVIEW

**Comment by JyeSmith** [comment]
@brandonrc and @CapnBry 

Tagging you guys because I have monkeyed with FM30 and Ghost targets regarding antenna switching.  Tx antenna will now ping pong with every packet in non DVDA mode.

**Comment by CapnBry** [comment]
I've read all the code, done all the consideration, I just haven't tried it and instrumented it to see how it works yet.

Quad send 250... what good is this if you could have dual send 500 and had a fresh channels packet instead? Is it strictly for Betaflight devs to have fun with and try new things? Is it for users who want to run 115200 handset baud?

**Comment by JyeSmith** [comment]
Testing is needed at large events where everyone leaves their handset on, and will help us see if there is any benefit of D250 over D500.  Park flying by yourself with no noise, D500 or straight up 1k might be fine.  But D250 may be beneficial at large events where LQ is getting crushed.

Ah... FML30.  I will revert with some comments.

**Comment by FishballFPV** [comment]
> testing is needed at large events where everyone leaves their handset on, and will help us see if there is any benefit of D250 over D500. Park flying by yourself with no noise, D500 or straight up 1k might be fine. But D250 may be beneficial at large events where LQ is getting crushed.
> 
Will it also make sense to use D250 when flying in a 2.4ghz congested area(city, housing with wifi being predominant) still trying to get my thoughts on the difference of D250, D500 and F1000 OTA, 

I guess the settings will definitely depend on the situation/environment im flying..

**Comment by JyeSmith** [comment]
@FishballFPV yep, the more noise the lower the rate you want to use.

FYI don't use 1k unless you have modified betaflight for a 1M baud.

**Comment by FishballFPV** [comment]
> @FishballFPV yep, the more noise the lower the rate you want to use.
> 
> FYI don't use 1k unless you have modified betaflight for a 1M baud.

got it..ill figure it out where exactly in betaflight i needed the 1M baud selected 🤣🤣.

**Comment by FishballFPV** [comment]
IMG20220505060357

i just noticed that the oled function isnt working after flashing dvda..thought i post it here..

**Comment by JyeSmith** [comment]
There might be an issue in the display libs due to the increase in rf modes.  Ill check it out.

**Comment by JyeSmith** [comment]
@FishballFPV the screen should be fixed.  Thanks for testing and let me of anything else.

**Comment by FishballFPV** [comment]
> @FishballFPV the screen should be fixed. Thanks for testing and let me of anything else.

thanks for the update,still stuck at elrs logo on the oled but everything works fine though. is it because the oled screen i use is just .96 thats why its not displaying everything?

**Comment by JyeSmith** [comment]
@FishballFPV I tested with the NamimnoRC which uses a SPI screen.  The Jumper Aion uses I2C like you are, so Ill test it when home.

**Comment by FishballFPV** [comment]
> @FishballFPV I tested with the NamimnoRC which uses a SPI screen. The Jumper Aion uses I2C like you are, so Ill test it when home.

ill try it again later,i used the same user define from v2.4

**Comment by FishballFPV** [comment]
> > @FishballFPV I tested with the NamimnoRC which uses a SPI screen. The Jumper Aion uses I2C like you are, so Ill test it when home.
> 
> ill try it again later,i used the same user define from v2.4


still stuck on elrs logo on the oled..🤓🤣
IMG20220507061821

**Comment by JyeSmith** [comment]
@FishballFPV fixed 😃 

There was already a fix on master and I have added it to this PR for your testing.

File: src/src/common.cpp [hunk 1]
```diff
@@ -29,20 +29,22 @@ expresslrs_rf_pref_params_s ExpressLRS_AirRateRFperf[RATE_MAX] = {
...
 
 expresslrs_mod_settings_s ExpressLRS_AirRateConfig[RATE_MAX] = {
-    {0, RADIO_TYPE_SX128x_FLRC, RATE_FLRC_1000HZ, SX1280_FLRC_BR_0_650_BW_0_6, SX1280_FLRC_BT_1, SX1280_FLRC_CR_1_2,     1000, TLM_RATIO_1_128,  8, 32, 8},
-    {1, RADIO_TYPE_SX128x_FLRC, RATE_FLRC_500HZ,  SX1280_FLRC_BR_0_650_BW_0_6, SX1280_FLRC_BT_1, SX1280_FLRC_CR_1_2,     2000, TLM_RATIO_1_128,  4, 32, 8},
-    {2, RADIO_TYPE_SX128x_LORA, RATE_LORA_500HZ,  SX1280_LORA_BW_0800,         SX1280_LORA_SF5,  SX1280_LORA_CR_LI_4_6,  2000, TLM_RATIO_1_128,  4, 12, 8},
-    {3, RADIO_TYPE_SX128x_LORA, RATE_LORA_250HZ,  SX1280_LORA_BW_0800,         SX1280_LORA_SF6,  SX1280_LORA_CR_LI_4_7,  4000, TLM_RATIO_1_64,   4, 14, 8},
-    {4, RADIO_TYPE_SX128x_LORA, RATE_LORA_150HZ,  SX1280_LORA_BW_0800,         SX1280_LORA_SF7,  SX1280_LORA_CR_LI_4_7,  6666, TLM_RATIO_1_32,   4, 12, 8},
-    {5, RADIO_TYPE_SX128x_LORA, RATE_LORA_50HZ,   SX1280_LORA_BW_0800,         SX1280_LORA_SF9,  SX1280_LORA_CR_LI_4_6, 20000, TLM_RATIO_NO_TLM, 2, 12, 8}};
+    {0, RADIO_TYPE_SX128x_FLRC, RATE_FLRC_1000HZ, SX1280_FLRC_BR_0_650_BW_0_6, SX1280_FLRC_BT_1, SX1280_FLRC_CR_1_2,     1000, TLM_RATIO_1_128,  2, 32, 8, 1},
```
Comment (pkendall64) [inline src/src/common.cpp:32]: Only question I have is why are we changing this one to only 2 packets on one frequency?

**Review by pkendall64 — APPROVED** [review]
Flew this today. Bloody awesome on the LQ!

File: src/src/common.cpp (hunk 1 above)
Comment (JyeSmith) [reply src/src/common.cpp:32]: On the scoreboard I see a lot blocks of interference with FLRC and not so much with lora.  But yeah, there's probably no reason not to change the other modes.

Comment (pkendall64) [reply src/src/common.cpp:32]: No that's fine, I was just wondering if it was a cut/paste artefact, thats all.

**Comment by FishballFPV** [comment]
@JyeSmith  so I theres a OpenI6X firmware release v1.5 for my FlyskyI6X and after the update ELRS Lua is working but on DVDA im unable to select both D250,D500 and F1000 and even the 500mhz packet rate. Im limited to 250hz max for some reason.

Just an educated guess since I heard that OpenI6X firmware are somewhat ELRS v3 ready? Fyi...OpenI6X v1.2 to 1.3 works fine with DVDA so I had to revert to OpenI6X v1.3 for now.  Not sure If I should also report it to the OpenI6X github.

**Comment by JyeSmith** [comment]
> @JyeSmith so there's a OpenI6X firmware release v1.5 for my FlyskyI6X and after the update ELRS Lua is working but on DVDA im unable to select either D250,D500 and F1000 and even the 500mhz packet rate. Im limited to 250hz max for some reason.
> 
> Just an educated guess since I heard that OpenI6X firmware are somewhat ELRS v3 ready? Fyi...OpenI6X v1.2 to 1.3 works fine with DVDA so I had to revert to OpenI6X v1.3 for now. Not sure If I should also report it to the OpenI6X github.

For anyone else that see this, ELRS does not allow setting RF modes greater than 250Hz while the baud is 115200.  The serial link between the handset and ELRS module is not fast enough.  Changing the baud to 400k or greater allows selection of the higher rates.

**Comment by AlessandroAU** [comment]
I thought there was a mistake with how the DVDA window was calculated on the RX side but looks like Jye was several steps ahead. PR looks good and is really quite elegant in implementation.

**Review by AlessandroAU — APPROVED** [review]

**Pull request merged**